docs/design/configs.md

```diff
@@ -0,0 +1,130 @@
+# Generic Config Object
+
+## Abstract
+
+A generic ``Config`` object to store command-line flags and environment variables
+that is used for configuration of ``Kubernetes`` components living on top of ``etcd``
+is proposed in this proposal.
+
+The main focus of this proposal to solve dynamic configuration problem of components 
+and encapsulate configuration information to increase overall system quality and to 
+create a flexible configuration model for ``Kubernetes`` components.
+
+## Problem
+
+Currently command-line flags and environment variables are used for component configurations in ``Kubernetes``.
+But, this approach is not suitable for a distributed computing environment. Specifically the 
+problems with the current approach are:
+
+* Hard to synchronize across replicated components.
+* Hard to change dynamically as needed by continuous services.
+* Hard to change in different versions that may support different configuration options.
+
+## Solution
+
+A new ``Config`` resource is proposed to solve the aforementioned problems 
+that is stored in ``etcd`` and distributed to interested/necessary components 
+through the API server. ``Config`` resource will be mounted as a new volume 
+and will be managed by a side-car container that communicates with the API server.
+
+The ``Config`` resource architecture will be very similar to that of ``Secrets``.
+However, the ``Config`` resource will be dynamic, meaning that different components consuming
+the same ``Config`` will be notified about the changes to that particular resource.
+
+### Advantages
+
+* Reusable across different components.
+* Easy distribution through the API server.
+* By leveraging the power of ``/watch API`` along with the new resource, we can change component configurations dynamically.
+* Layer of abstraction that gets rid of the global state currently used for command-line flags and environment variables.
+
+### API Resource
+
+A new resource for ``Config`` will be added to the ``API``:
+
+```go
+	type Config struct {
+		TypeMeta   `json:",inline"`
+		ObjectMeta `json:"metadata,omitempty"`
+
+		// Flags contains command-line flags.
+		Flags pflag.FlagSet `json:"flags,omitempty"`
```

Comment:
Dynamic is probably a case for eventual consistency a la downward api
volume.

For static we should be able to put these into envs w/o a lot of risk but
will have to validate size.
On Sat, Jun 20, 2015 at 5:24 PM Clayton Coleman <notifications@github.com>
wrote:

> In docs/design/configs.md
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6477#discussion_r32888204>
> :
>
> > +* Reusable across different components.
> > +* Easy distribution through the API server.
> > +* By leveraging the power of ``/watch API`` along with the new resource, we can change component configurations dynamically.
> > +* Layer of abstraction that gets rid of the global state currently used for command-line flags and environment variables.
> > +
> > +### API Resource
> > +
> > +A new resource for ``Config`` will be added to the ``API``:
> > +
> > +```go
> > +	type Config struct {
> > +		TypeMeta   `json:",inline"`
> > +		ObjectMeta `json:"metadata,omitempty"`
> > +
> > +		// Flags contains command-line flags.
> > +		Flags pflag.FlagSet `json:"flags,omitempty"`
>
> Would much prefer map[string]string, and an optional type, like secrets.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6477/files#r32888204>
> .
>